okhttp/src/main/java/okhttp3/internal/connection/Transmitter.java

```diff
@@ -303,7 +305,7 @@ public void exchangeDoneDueToException() {
   }
 
   public boolean canRetry() {
-    return exchangeFinder.canRetry();
+    return exchangeFinder.hasStreamFailure() && exchangeFinder.hasRouteToTry();
```

Comment:
Is it problematic that this locks twice allowing code to run between the two expressions?

Reply:
Nope. Whew!